src/object.c

```diff
@@ -211,6 +211,304 @@ robj *dupStringObject(const robj *o) {
+        streamCG *new_cg = streamCreateCG(new_s, (char *)ri_cgroups.key,
+                                          ri_cgroups.key_len, &cg->last_id);
+
+        serverAssert(new_cg != NULL);
+
+        /* Consumers */
+        raxIterator ri_consumers;
+        raxStart(&ri_consumers, cg->consumers);
+        raxSeek(&ri_consumers, "^", NULL, 0);
+        while (raxNext(&ri_consumers)) {
+            streamConsumer *consumer = ri_consumers.data;
+            streamConsumer *new_consumer =
+                raxFind(new_cg->consumers, (unsigned char *)consumer->name,
+                        sdslen(consumer->name));
+            if (new_consumer == raxNotFound) {
+                new_consumer = zmalloc(sizeof(*new_consumer));
+                new_consumer->name = sdsdup(consumer->name);
+                new_consumer->pel = raxNew();
+                raxInsert(new_cg->consumers,
+                          (unsigned char *)new_consumer->name,
+                          sdslen(new_consumer->name), new_consumer, NULL);
+                new_consumer->seen_time = consumer->seen_time;
+            }
+
+            /* Consumer PEL */
+            raxIterator ri_cpel;
+            raxStart(&ri_cpel, consumer->pel);
+            raxSeek(&ri_cpel, "^", NULL, 0);
+            while (raxNext(&ri_cpel)) {
+                unsigned char buf[sizeof(streamID)];
+                streamNACK *nack = ri_cpel.data;
+                streamDecodeID(ri_cpel.key, &id);
+                streamEncodeID(buf, &id);
+                /* Insert NACK. */
+                streamNACK *new_nack = zmalloc(sizeof(*new_nack));
+                new_nack->delivery_time = nack->delivery_time;
+                new_nack->delivery_count = nack->delivery_count;
+                new_nack->consumer = new_consumer;
+                int group_inserted =
+                    raxTryInsert(new_cg->pel, buf, sizeof(buf), new_nack, NULL);
+                int consumer_inserted = raxTryInsert(
+                    new_consumer->pel, buf, sizeof(buf), new_nack, NULL);
+
+                if (group_inserted == 0) {
+                    streamFreeNACK(new_nack);
+                    new_nack = raxFind(new_cg->pel, buf, sizeof(buf));
+                    serverAssert(new_nack != raxNotFound);
+                    raxRemove(new_nack->consumer->pel, buf, sizeof(buf), NULL);
+                    /* Update the consumer and NACK metadata. */
+                    new_nack->delivery_time = nack->delivery_time;
+                    new_nack->delivery_count = nack->delivery_count;
+                    new_nack->consumer = new_consumer;
+                    /* Add the entry in the new consumer local PEL. */
+                    raxInsert(new_consumer->pel, buf, sizeof(buf), new_nack,
+                              NULL);
+                } else if (group_inserted == 1 && consumer_inserted == 0) {
+                    serverPanic("NACK half-created. Should not be possible.");
+                }
```

Comment:
IIUC what this code does it:
1. create a NACK
2. insert it to the group and the consumer
3. if the group insertion failed (since the group already has one):
	1. delete the newly created nack
	2. locate the the group nack
	3. remove the newly added consumer nack
	4. add the group nack into the consumer.

to me this seems like a complicated and wrong approach, the more trivial one IMHO is:
1. search the group, if it already has one:
	1. add it to the consumer
2. else:
	1. create a new nack
	3. add to both the group and consumer

advantage:
- more logical approach IMHO (and probably shorter too)
- not allocating a nack just in order to free it soon.
- involves 3 rax searches max (1 find and 2 insertions), existing approach has 5 (2 insertions, find, removal and insertion)

Reply:
looking at your recent change, i don't think that's good.
the PEL record in the consumer is expected to be the same pointer as the one in the consumer group PEL.
i see your recent code just calls `raxTryInsert` and `raxInsert`, so if the same record exists in two consumers, one will be shared with the group, and the other one will not (memory leak later).

however, i see that raxTryInsert and raxInsert can return the existing pointer if there's a clash.
so you can use that.
this will mean that you'll have just two searches (not 3 as in my previous proposal), but only other hand you'll have excessive calls to malloc / free, i suppose that's not so bad, and the code will be cleaner.

Reply:
thanks @oranagra
I'm sorry but, I didn't understand exactly how to fix it, so I took a look at rdb.c to fix it.
Would this fix be a problem?

Reply:
@swamp0407 new code looks good. (being the same approach that rdb loading takes, it can't be wrong)
also, it seems i got confused, one nack record can't exist in two consumers.